File: backend-spring/src/main/java/com/cronos/service/UserService.java
package com.cronos.service;

import com.cronos.dto.request.CreateUserRequest;
import com.cronos.dto.response.PaginationMeta;
import com.cronos.dto.response.UserResponse;
import com.cronos.entity.User;
import com.cronos.repository.UserRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Sort;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class UserService {

    private final UserRepository userRepo;
    private final PasswordEncoder passwordEncoder;

    public Map<String, Object> getAllUsers(int page, int limit, String role, String search) {
        Page<User> usersPage = userRepo.findAllFiltered(
                role, search,
                PageRequest.of(page - 1, limit, Sort.by(Sort.Direction.DESC, "createdAt")));
        List<UserResponse> responses = usersPage.getContent().stream()
                .map(UserResponse::fromEntity).collect(Collectors.toList());
        PaginationMeta meta = PaginationMeta.builder()
                .currentPage(page).perPage(limit)
                .total(usersPage.getTotalElements())
                .totalPages(usersPage.getTotalPages()).build();
        return Map.of("users", responses, "meta", meta);
    }

    public UserResponse createUser(CreateUserRequest req) {
        if (userRepo.findByEmail(req.getEmail()).isPresent()) {
            throw new RuntimeException("email already registered");
        }
        User user = User.builder()
                .name(req.getName()).email(req.getEmail())
                .password(passwordEncoder.encode(req.getPassword()))
                .phone(req.getPhone()).role(req.getRole())
                .isVerified(true).balance(BigDecimal.ZERO).build();
        userRepo.save(user);
        return UserResponse.fromEntity(user);
    }

    public UserResponse verifyUser(UUID userId) {
        User user = userRepo.findById(userId).orElseThrow(() -> new RuntimeException("user not found"));
        if (!"petambak".equals(user.getRole()) && !"logistik".equals(user.getRole())) {
            throw new RuntimeException("only petambak and logistik accounts need verification");
        }
        user.setIsVerified(true);
        userRepo.save(user);
        return UserResponse.fromEntity(user);
    }

    public UserResponse updateUserStatus(UUID userId, boolean isVerified) {
        User user = userRepo.findById(userId).orElseThrow(() -> new RuntimeException("user not found"));
        user.setIsVerified(isVerified);
        userRepo.save(user);
        return UserResponse.fromEntity(user);
    }

    public int[] parsePagination(String pageStr, String limitStr) {
        int page = 1, limit = 10;
        try { page = Integer.parseInt(pageStr); } catch (Exception ignored) {}
        try { limit = Integer.parseInt(limitStr); } catch (Exception ignored) {}
        if (page < 1) page = 1;
        if (limit < 1) limit = 10;
        if (limit > 100) limit = 100;
        return new int[]{page, limit};
    }
}

File: backend/services/user_service.go
package services
import (
	"errors"
	"math"
	"strconv"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"cronos-backend/utils"
	"github.com/google/uuid"
)
type UserService struct {
	UserRepo *repositories.UserRepository
}
func NewUserService(userRepo *repositories.UserRepository) *UserService {
	return &UserService{UserRepo: userRepo}
}
func (s *UserService) GetAllUsers(page, limit int, role, search string) ([]models.UserResponse, *utils.PaginationMeta, error) {
	users, total, err := s.UserRepo.FindAll(page, limit, role, search)
	if err != nil {
		return nil, nil, errors.New("failed to fetch users")
	}
	var responses []models.UserResponse
	for _, u := range users {
		responses = append(responses, u.ToResponse())
	}
	meta := &utils.PaginationMeta{
		CurrentPage: page,
		PerPage:     limit,
		Total:       total,
		TotalPages:  int(math.Ceil(float64(total) / float64(limit))),
	}
	return responses, meta, nil
}
func (s *UserService) VerifyUser(userID uuid.UUID) (*models.UserResponse, error) {
	user, err := s.UserRepo.FindByID(userID)
	if err != nil {
		return nil, errors.New("user not found")
	}
	if user.Role != "petambak" && user.Role != "logistik" {
		return nil, errors.New("only petambak and logistik accounts need verification")
	}
	user.IsVerified = true
	if err := s.UserRepo.Update(user); err != nil {
		return nil, errors.New("failed to verify user")
	}
	resp := user.ToResponse()
	return &resp, nil
}
func (s *UserService) CreateUser(req models.CreateUserRequest) (*models.UserResponse, error) {
	existing, _ := s.UserRepo.FindByEmail(req.Email)
	if existing != nil && existing.ID != uuid.Nil {
		return nil, errors.New("email already registered")
	}
	hashedPassword, err := utils.HashPassword(req.Password)
	if err != nil {
		return nil, errors.New("failed to hash password")
	}
	user := &models.User{
		ID:         uuid.New(),
		Name:       req.Name,
		Email:      req.Email,
		Password:   hashedPassword,
		Phone:      req.Phone,
		Role:       req.Role,
		IsVerified: true, 
	}
	if err := s.UserRepo.Create(user); err != nil {
		return nil, errors.New("failed to create user")
	}
	resp := user.ToResponse()
	return &resp, nil
}
func (s *UserService) UpdateUserStatus(userID uuid.UUID, isVerified bool) (*models.UserResponse, error) {
	user, err := s.UserRepo.FindByID(userID)
	if err != nil {
		return nil, errors.New("user not found")
	}
	user.IsVerified = isVerified
	if err := s.UserRepo.Update(user); err != nil {
		return nil, errors.New("failed to update user status")
	}
	resp := user.ToResponse()
	return &resp, nil
}
func (s *UserService) ParsePagination(pageStr, limitStr string) (int, int) {
	page, err := strconv.Atoi(pageStr)
	if err != nil || page < 1 {
		page = 1
	}
	limit, err := strconv.Atoi(limitStr)
	if err != nil || limit < 1 {
		limit = 10
	}
	if limit > 100 {
		limit = 100
	}
	return page, limit
}
